sympy/matrices/random.py

```diff
@@ -0,0 +1,2248 @@
+import random
+
+from ..core import I as _i
+from ..core.mul import Mul as _multiply
+from ..functions import sqrt as _sqrt, re as _re, im as _im, \
+    transpose as _t, adjoint as _c
+from ..tensor import shape as _shape
+from .dense import eye as _eye
+
+__all__ = 'identity_matrix', 'projection', 'jordan', 'transposition', \
+          'permutation', 'elementary', 'rotation', 'reflection', \
+          'diagonal_normal', 'jordan_normal', 'isometry_normal', \
+          'triangular', 'invertible', 'singular', \
+          'idempotent', 'nilpotent', 'diagonalizable', 'trigonalizable', \
+          'orthogonal', 'unitary', 'normal', \
+          'symmetric', 'hermite', \
+          'regular_to_singular', 'complex_to_real'
+
+_TEST = False
+
+_eps = 1e-13
+
+
+def _get_value(o, i, j):
+    return o[i, j]
+
+
+def _set_value(o, i, j, v):
+    o[i, j] = v
+
+
+def _inv(o):
+    return o.inv()
+
+
+def _cs(scalar):
+    sgn = random.choice((-1, 1))
+    if isinstance(scalar, (tuple, list)) and len(scalar) == 2:
+        return scalar
+    if isinstance(scalar, complex) or getattr(scalar, 'is_real', True) is False:
+        return _re(scalar), _im(scalar)
+    c = scalar
+    s = sgn * _sqrt(1 - c * c)
+    return c, s
```

Comment:
What exactly does this function do? Is it returning a random cos/sin pair?

Reply:
This functions provides proper eigenvalues of isometries. In latest version it can have real output (cos/sin) or complex (cos + sin *I). For convenience, different input types are handled (cos/sin, complex units or just cos (with random sin)). It ensures same behaviour for all isometry related matrices (like `rotation`, `reflection`, `isometry_normal` (form), `orthogonal`, `unitary` and `normal`).